core/src/main/scala/kafka/cluster/Partition.scala

```diff
@@ -202,6 +202,13 @@ class Partition(val topic: String,
           true
         }
       val leaderReplica = getReplica().get
+      val curLeaderLogEndOffset = leaderReplica.logEndOffset.messageOffset
+      val curTimeMs = time.milliseconds
+      // initialize lastCaughtUpTime of replicas as well as their lastFetchTimeMs and lastFetchLeaderLogEndOffset.
+      (assignedReplicas() - leaderReplica).foreach{replica =>
+        val lastCaughtUpTimeMs = if (inSyncReplicas.contains(replica)) curTimeMs else 0L
+        replica.resetLastCaughtUpTime(curLeaderLogEndOffset, curTimeMs, lastCaughtUpTimeMs)
+      }
```

Comment:
A couple of things.
1. I think we only need to call resetLastCaughtUpTime() if isNewLeader is true.
2. It's a bit confusing to explicitly pass in lastFetchLeaderLogEndOffset and lastFetchTimeMs. I am thinking perhaps it's better to have two separate methods resetAsCaughtUp(currentTime: Long) and resetAsNotCaughtUp(). For in-sync replicas, we call the former and for other replicas, we call the latter. In resetAsCaughtUp(), we set lastCaughtUpTime to currentTime, and also set lastFetchLeaderLogEndOffset = 0 and lastFetchTimeMs = currentTime, which makes it a bit hard for the replica to get dropped out of ISR. In resetAsNotCaughtUp(), we set the lastCaughtUpTime to 0 and also set lastFetchLeaderLogEndOffset = MAX_LONG and lastFetchTimeMs = 0, which makes it a bit hard for the replica to get added back to ISR.